providers/ldap/ldap_provider.go

```diff
@@ -0,0 +1,624 @@
+package ldap
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher/v2"
+	"github.com/rancher/rancher-auth-service/model"
+	"gopkg.in/ldap.v2"
+)
+
+type SearchConfig struct {
+	Server               string
+	Port                 int64
+	BindDN               string
+	BindPassword         string
+	UserSearchAttributes []string
+	GroupSeachAttributes []string
+}
+
+var searchConfig *SearchConfig
+var nilIdentity = client.Identity{Resource: client.Resource{
+	Type: "identity",
+}}
+var nilToken = model.Token{Resource: client.Resource{
+	Type: "token",
+}}
+
+func InitializeSearchConfig(l *LdapProvider) *SearchConfig {
+	return &SearchConfig{
+		Server:       l.ldapClient.config.Server,
+		Port:         l.ldapClient.config.Port,
+		BindDN:       l.ldapClient.config.ServiceAccountUsername,
+		BindPassword: l.ldapClient.config.ServiceAccountPassword,
+		UserSearchAttributes: []string{l.ConstantsConfig.memberOfAttribute,
+			l.ConstantsConfig.objectClassAttribute,
+			l.ldapClient.config.UserObjectClass,
+			l.ldapClient.config.UserLoginField,
+			l.ldapClient.config.UserNameField,
+			l.ldapClient.config.UserSearchField,
+			l.ldapClient.config.UserEnabledAttribute},
+		GroupSeachAttributes: []string{l.ConstantsConfig.memberOfAttribute,
+			l.ConstantsConfig.objectClassAttribute,
+			l.ldapClient.config.GroupObjectClass,
+			l.ldapClient.config.UserLoginField,
+			l.ldapClient.config.GroupNameField,
+			l.ldapClient.config.GroupSearchField},
+	}
+}
+
+// GenerateToken generates token
+func GenerateToken(l *LdapProvider, jsonInput map[string]string) (model.Token, error) {
+	log.Info("Now generating Ldap token")
+	searchConfig = InitializeSearchConfig(l)
+	split := strings.Split(jsonInput["code"], ":")
+	username, password := split[0], split[1]
+	externalId := username
+	if l.ldapClient.config.LoginDomain != "" {
+		externalId = l.ldapClient.config.LoginDomain + "\\" + username
+	}
+
+	if password == "" {
+		return nilToken, fmt.Errorf("Failed to login, password not provided")
+	}
+
+	log.Info("Now creating Ldap connection")
+	lConn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", searchConfig.Server, searchConfig.Port))
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer lConn.Close()
+
+	log.Info("Binding username password")
+	err = lConn.Bind(externalId, password)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
This should be logged at Error, not Fatal - fatal will quit the rancher-auth-service. Also you need to return the Errors back to the caller, so return the error by using  fmt.Errorf("",err) and it should get logged at the origin of the call.